# ANA540 Midterm Exam_Summer

**Instructions:**

- **Due by**: Monday 06/16 by 8 a. m. (E.T.)

- Please submit your work as a .ipynb notebook or a PDF file on Blackboard.

- Please include your name in the filename of your submission.

<br>

## Part 1： Multiple choice question and True/False Questions
(2pts/question, total 30 points)

#### 1. Which of the following best describes machine learning?

a) A method for humans to code specific instructions for every task.

b) A method of teaching computers to make decisions based on data.

c) A method to improve computer memory.

d) A tool for data storage.


Answer: b


#### 2. In supervised learning, what do we provide the algorithm with?

a) Only the input data.

b) Only the output data.

c) Both input data and the corresponding output.

d) Neither input nor output data.



Answer: c

#### 3. Which of the following best describes the KNN algorithm?

a) An algorithm that clusters data into K groups based on distance.

b) An algorithm that predicts the output of a new data point by averaging the outputs of its K nearest neighbors.

c) An algorithm that fits a linear boundary to separate data into K classes

d) An algorithm that trains on K random subsets of the data to make predictions.




Answer: b

#### 4. In linear regression, what is the purpose of the cost function?

a) To minimize the difference between predicted and actual values

b) To maximize the number of features used

c) To find the number of clusters in the data

d) To reduce the dimensionality of the dataset


Answer: a

#### 5. Which of the following is TRUE about gradient descent?

a) It always converges to the global minimum.

b) It can get stuck in local minima.

c) It does not require a learning rate.

d) It is not used for optimization problems.


Answer: b

#### 6. In a Naive Bayes classifier, the 'naive' assumption is that:

a) All features are unrelated to the output.

b) All features are related to each other.

c) All features are conditionally independent given the output.

d) The output is independent of all features.



Answer: c

#### 7. Which statements about features and labels are true in a regression problem? (Multiple answers may be correct.)

a) Features are always numeric, while labels are categorical.

b) Features and labels are both always categorical.

c) Labels are the variables we are trying to predict, and they can be numeric.

d) Features are the inputs, and labels are the outputs.


Answer: c, d

#### 8. Which of the following best describes the main objective of Linear SVM?

a) To cluster similar data points together.

b) To minimize the sum of squared errors.

c) To find the hyperplane that best separates the classes with the maximum margin.

d) To probabilistically predict class membership.


Answer: c

#### 9. What does the term "margin" refer to in the context of linear support vector machines?

a) The distance from the support vectors to the separating hyperplane.

b) The shortest distance from the positive and negative samples to the separating hyperplane.

c) The longest distance from the positive and negative samples to the separating hyperplane.

d) Both A and B are correct.

Answer: d

#### 10. In which scenario is feature scaling particularly important? (Multiple answers may be correct.)

a) When working with algorithms that rely on distances, such as KNN.

b) When using gradient descent optimization algorithms.

c) When the dataset consists only of categorical variables.

d) When all features are already in the same range and scale.

Answer: a b

#### 11. Logistic regression can only be used for binary classification problems.

True or False?



Answer: False

#### 12. In the k-Nearest Neighbors (k-NN) algorithm, choosing a very large value of k will always improve the model's performance.

True or False?

Answer: False

#### 13. Machine learning models can only be used to make predictions on new, unseen data.

True or False?

Answer: False

#### 14. For better model accuracy, training data should overlap with testing data to ensure the model performs well on known examples.

True or False?

Answer: False

#### 15. In machine learning, increasing the amount of training data alone can solve the problem of overfitting.

True or False?

Answer: False

## Part 2: Case Study 

(5 points)

#### Case Study: Predicting Customer Churn for a Telecommunications Company 

*Scenario*:
As a data scientist at a telecommunications company, you are tasked with developing a machine learning model to predict customer churn. Your dataset includes features such as customer demographics (age, gender), service usage (call duration, internet usage), contract information (contract type, tenure), and billing details (monthly charges, payment method). The label indicates whether a customer has churned or not.



**a) Select and briefly describe two (or more) machine learning algorithm suitable for this problem, and explain your reasoning for choosing it.**

Firstly, this problem can be categorized as a binary classification problem. By using labels, we can train a model to predict which customer will be churn. 
For binary classification model, we can use Logistic Regression, Decision Tree, Random Forest. Those models are good for this problem because they can handle classification and good explainability.
Regarding Logistic Regression, it is a simple and fast model, but it may not be accurate enough because it requires linear relationship between features and target variable. So I will use Decision Tree or Random Forest.



**b) Briefly, what issues should be considered during the data preprocessing phase, and what actions need to be taken?**

Couple potential issues on this dataset. We need to work on data scaling like normalization or standardization beacause each numeric features are different unit(age, call duration, etc.) larger values may give more noise than smaller values. Also, we need to work on one-hot encoding for categorical features. It's important to drop one category from each set of encoded features to avoid the dummy variable trap. We should also carefully check the data for any missing values or inconsistencies, such as negative values in tenure or charges, and handle these appropriately.



<br>

## Part 3: Feature Scaling and Classification with Logistic Regression and Linear SVM using Wine Dataset 

(65 points)

### Wine Dataset Description:
The Wine dataset is a built-in dataset in Scikit-Learn primarily used for classification tasks. It originates from three different types of wine, each described by the results of 13 different chemical analyses. The objective is to predict the origin of the wine based on these chemical attributes.

### Feature Explanations:
- alcohol - Alcohol content
- malic_acid - Malic acid content, a naturally occurring organic acid that is found in various plants and fruits.
- ash - Total ash content, a measure of the inorganic residue remaining after the evaporation of water and the removal of organic material.
- alcalinity_of_ash - Alkalinity of the ash, measuring how basic the ash is.
- magnesium - Magnesium content.
- total_phenols - Total phenols content, compounds which have antioxidant properties.
- flavanoids - Flavanoid content, a group of polyphenols which are known for their health benefits.
- nonflavanoid_phenols - Non-flavanoid phenols content, another type of polyphenol but not classified as flavanoids.
- proanthocyanins - Proanthocyanidin content, polyphenolic antioxidants found in many plants.
- color_intensity - Color intensity, an indicative measure of the wine type and quality.
- hue - Hue, the type of color the wine has.
- od280/od315_of_diluted_wines - Ratio of OD280 to OD315 of diluted wines, a measure often used to gauge the protein content in wine.
- proline - Proline content, an amino acid that is often used as a marker for wine quality.


    - The target variable, named **target**, indicates the origin of the wine, with three possible classes.


In [44]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.datasets import load_wine
import warnings
warnings.filterwarnings("ignore")

from sklearn.datasets import load_wine

# Load the Wine dataset
wine = load_wine()

# Convert the Wine dataset to a Pandas DataFrame
df = pd.DataFrame(wine.data, columns=wine.feature_names)
df['target'] = wine.target
df.head(10)

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
0,14.23,1.71,2.43,15.6,127.0,2.80,3.06,0.28,2.29,5.64,1.04,3.92,1065.0,0
1,13.20,1.78,2.14,11.2,100.0,2.65,2.76,0.26,1.28,4.38,1.05,3.40,1050.0,0
2,13.16,2.36,2.67,18.6,101.0,2.80,3.24,0.30,2.81,5.68,1.03,3.17,1185.0,0
3,14.37,1.95,2.50,16.8,113.0,3.85,3.49,0.24,2.18,7.80,0.86,3.45,1480.0,0
4,13.24,2.59,2.87,21.0,118.0,2.80,2.69,0.39,1.82,4.32,1.04,2.93,735.0,0
5,14.20,1.76,2.45,15.2,112.0,3.27,3.39,0.34,1.97,6.75,1.05,2.85,1450.0,0
6,14.39,1.87,2.45,14.6,96.0,2.50,2.52,0.30,1.98,5.25,1.02,3.58,1290.0,0
7,14.06,2.15,2.61,17.6,121.0,2.60,2.51,0.31,1.25,5.05,1.06,3.58,1295.0,0
8,14.83,1.64,2.17,14.0,97.0,2.80,2.98,0.29,1.98,5.20,1.08,2.85,1045.0,0
9,13.86,1.35,2.27,16.0,98.0,2.98,3.15,0.22,1.85,7.22,1.01,3.55,1045.0,0


<br>

### Q1. Check some basic data information
(10 pts)

1. Sample Size: Check number of rows and columns

2. Data Type: Check data types of columns

3. Class Balance: Check the distribution of the Target variable

4. Check summary statistics: Review the mean, median, standard deviation, range of each feature, etc,. This helps us understand the data distribution and decide whether feature scaling is necessary.

5. Missing Values: Check for missing values

1. Sample size:

In [45]:

df.shape

(178, 14)

2. Data Type

In [46]:
df.dtypes

alcohol                         float64
malic_acid                      float64
ash                             float64
alcalinity_of_ash               float64
magnesium                       float64
total_phenols                   float64
flavanoids                      float64
nonflavanoid_phenols            float64
proanthocyanins                 float64
color_intensity                 float64
hue                             float64
od280/od315_of_diluted_wines    float64
proline                         float64
target                            int64
dtype: object

3. Class balance

In [47]:
df['target'].value_counts()


target
1    71
0    59
2    48
Name: count, dtype: int64

4. Summary Statistics


In [48]:
df.describe()

,alcohol,malic_acid,ash,alcalinity_of_ash,magnesium,total_phenols,flavanoids,nonflavanoid_phenols,proanthocyanins,color_intensity,hue,od280/od315_of_diluted_wines,proline,target
count,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000,178.000000
mean,13.000618,2.336348,2.366517,19.494944,99.741573,2.295112,2.029270,0.361854,1.590899,5.058090,0.957449,2.611685,746.893258,0.938202
std,0.811827,1.117146,0.274344,3.339564,14.282484,0.625851,0.998859,0.124453,0.572359,2.318286,0.228572,0.709990,314.907474,0.775035
min,11.030000,0.740000,1.360000,10.600000,70.000000,0.980000,0.340000,0.130000,0.410000,1.280000,0.480000,1.270000,278.000000,0.000000
25%,12.362500,1.602500,2.210000,17.200000,88.000000,1.742500,1.205000,0.270000,1.250000,3.220000,0.782500,1.937500,500.500000,0.000000
50%,13.050000,1.865000,2.360000,19.500000,98.000000,2.355000,2.135000,0.340000,1.555000,4.690000,0.965000,2.780000,673.500000,1.000000
75%,13.677500,3.082500,2.557500,21.500000,107.000000,2.800000,2.875000,0.437500,1.950000,6.200000,1.120000,3.170000,985.000000,2.000000
max,14.830000,5.800000,3.230000,30.000000,162.000000,3.880000,5.080000,0.660000,3.580000,13.000000,1.710000,4.000000,1680.000000,2.000000


<br>

5. Missing values

In [49]:
df.isnull().sum()

alcohol                         0
malic_acid                      0
ash                             0
alcalinity_of_ash               0
magnesium                       0
total_phenols                   0
flavanoids                      0
nonflavanoid_phenols            0
proanthocyanins                 0
color_intensity                 0
hue                             0
od280/od315_of_diluted_wines    0
proline                         0
target                          0
dtype: int64

### Q2: Data Splitting and Preparation
(5 pts)

Split the Wine dataset into training and test sets and inspect the size of both sets.

-  Allocate 80% of the data for training and 20% for testing.
-  Use `random_state = 7`
-  Display the size of the training and testing data

In [50]:
from sklearn.model_selection import train_test_split

X = df.drop('target', axis=1)
y = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)


In [51]:
print("x_training size:", X_train.shape)
print("y_training size:", y_train.shape)
print("x_test size:", X_test.shape)
print("y_test size:", y_test.shape)

x_training size: (142, 13)
y_training size: (142,)
x_test size: (36, 13)
y_test size: (36,)


### Q3: Feature Scaling

(10 pts)

Standardize the features in the training and test sets.

- Use `StandardScaler` from scikit-learn.
- Fit the scaler on the training data and transform both the training and testing data.
- Display the first 5 rows of the scaled training data.

*Note*: All subsequent model building should use the scaled features (e.g., `X_train_scaled` for training and `X_test_scaled` for testing).

In [52]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = pd.DataFrame(scaler.transform(X_train), columns=X_train.columns)
X_test_scaled = pd.DataFrame(scaler.transform(X_test), columns=X_test.columns)

In [53]:

print("X_train_scaled:")
print(X_train_scaled.head(5))


print("X_test_scaled:")
print(X_test_scaled.head(5))



X_train_scaled:
    alcohol  malic_acid       ash  alcalinity_of_ash  magnesium  \
0 -0.109781    0.986826 -0.055207          -0.270719   0.400136   
1  0.825554    2.939229  0.299518           0.315289  -0.366751   
2  1.336870   -0.761462 -0.161625          -0.768827  -0.366751   
3  0.663429    0.214739  1.150857           1.487306   0.330419   
4  0.314237   -0.548473 -0.800129          -0.710226  -0.436468   

   total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
0      -1.424148   -1.348592              0.315056        -1.130436   
1      -0.974682   -1.437680              1.272017        -0.922865   
2      -0.168741    0.363881             -0.801398        -0.040686   
3      -1.176167   -1.210010              0.235309        -0.092579   
4       0.141236    0.126313             -0.721652        -0.421234   

   color_intensity       hue  od280/od315_of_diluted_wines   proline  
0         0.070159 -1.273811                     -1.236580 -0.299237  
1         1.

<br>

### Q4: Logistic regression model
(20 pts)

##### a)
- Instantiate a logistic regression model.
- Train the model on the training data.
- Make predictions on the training set and calculate the accuracy (optional).
- Make predictions on the test set and calculate the accuracy.

In [54]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

logisticRegression = LogisticRegression(random_state=7)
logisticRegression.fit(X_train, y_train)

# prediction against training data
y_pred_train = logisticRegression.predict(X_train)
print("Accuracy: ", accuracy_score(y_train, y_pred_train))


# prediction against test data
y_pred_test = logisticRegression.predict(X_test)
print("Accuracy: ", accuracy_score(y_test, y_pred_test))


Accuracy:  0.971830985915493
Accuracy:  0.9722222222222222


##### b)：Probability Predictions with Logistic Regression

- Predict probabilities for the test set using the trained logistic regression model
- Display the first 7 rows of the predicted probabilities.

Hint: Use the `predict_proba` method.

In [55]:
y_test_proba = logisticRegression.predict_proba(X_test_scaled)

print("Predicted probabilities for the first 7 samples:")
proba_df = pd.DataFrame(y_test_proba[:7], columns=['Class_0', 'Class_1', 'Class_2'])
print(proba_df)

Predicted probabilities for the first 7 samples:
    Class_0   Class_1   Class_2
0  0.000817  0.000020  0.999163
1  0.419355  0.548084  0.032560
2  0.000282  0.000006  0.999712
3  0.000139  0.000007  0.999854
4  0.341088  0.638196  0.020715
5  0.003196  0.000859  0.995945
6  0.398650  0.543930  0.057420


**Question: If a sample has a predicted probability of 98.66% for class_0, 1.20% for class_1, and 0.14% for class_2, how would you interpret the model's prediction for this sample?**

Answer: The model is extremely confident that sample belong to class_0(98.66%), then there's a very small chance the sample could be class_1(1.34%), and the model is less confident about class_2(0.14%).


Practically, this is a strong, confident prediction for class_0. The model sees very little evidence for class_1 or class_2, indicating that this sample is most likely to belong to class_0.


<br>

### Q5: Linear Support Vector Machine (SVM) model
(20 pts)

#### a):
- Instantiate a linear SVM model.
- Train the model on training data.
- Calculate accuracy on training and test set.

In [56]:
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score  

svm = LinearSVC(random_state=7)
svm.fit(X_train_scaled, y_train)

y_pred_train = svm.predict(X_train_scaled)
print("Training Accuracy: ", accuracy_score(y_train, y_pred_train))


y_pred_test = svm.predict(X_test_scaled)
print("Test Accuracy: ", accuracy_score(y_test, y_pred_test))



Training Accuracy:  1.0
Test Accuracy:  0.9444444444444444


#### b). Exploring the Impact of Regularization Parameter C:

- To examine the impact of the regularization strength on the model's training and test accuracy.
- Understand how the choice of C influences the model's generalization capability.

Train a Linear SVM model with different values of C: 0.001, 0.01, 0.1, 1, 10, and 100. Show the training and test accuracies for each value.

In [57]:
c_values = [0.001, 0.01, 0.1, 1, 10, 100]

results = []

for c in c_values:
    svm = LinearSVC(C=c, random_state = 7)
    svm.fit(X_train_scaled, y_train)

    y_pred_test = svm.predict(X_test_scaled)
    y_pred_train = svm.predict(X_train_scaled)

    test_accuracy = accuracy_score(y_test, y_pred_test)
    train_accuracy = accuracy_score(y_train, y_pred_train)
   
    results.append({
        'C': c,
        'Test Accuracy': test_accuracy,
        'Train Accuracy': train_accuracy
    })
    

results_df = pd.DataFrame(results)
print(results_df)


         C  Test Accuracy  Train Accuracy
0    0.001       1.000000        0.971831
1    0.010       1.000000        0.992958
2    0.100       0.972222        1.000000
3    1.000       0.944444        1.000000
4   10.000       0.944444        1.000000
5  100.000       0.944444        1.000000


**Question: How would you select an optimal value for C based on the training and test accuracies?**

Answer: 
Regarding model evalutation, We should pick up the model with the highest Test accuracy. 
Given that rule, C value 0.001 and 0.01 are the best model among different C values. Smaller C values regularize the model, and larger C values do not regularize the model, so if test accuracy is same, we should pick up the model with smaller C value. Hence, C value 0.001 is the best model. 


<br>